src/chain/chain.rs

```diff
@@ -0,0 +1,331 @@
+use std::borrow::Borrow;
+use std::marker::PhantomData;
+
+use hal::pso::PipelineStage;
+
+use chain::pass::PassLinks;
+use chain::link::{Acquire, Link, LinkSync, Release};
+use resource::{Access, Layout, Resource, Usage};
+
+/// Unique identifier for resource dependency chain.
+/// Multiple resource can be associated with single chain
+/// if all passes uses them the same way.
+/// Chain id uses marker type so that ids for buffers and images are different.
+#[derive(Derivative)]
+#[derivative(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct ChainId<T: ?Sized>(usize, PhantomData<T>);
+impl<T> ChainId<T> {
+    /// Make new chain id.
+    pub fn new(index: usize) -> Self {
+        ChainId(index, PhantomData)
+    }
+
+    /// Get index value.
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Resource chain for set of resources.
```

Comment:
similarly, use more prepositions, e.g. "Resource chain for **a** set of resources"